### PR TITLE
types: fix fsp for datetime, time and timestamp

### LINKED ISSUE #27159: got wrong result while use expression index
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
CREATE TABLE `t3` (
  `a` date DEFAULT NULL,
  `b` datetime DEFAULT NULL,
  `c` time DEFAULT NULL,
  `d` timestamp NULL DEFAULT NULL,
  `e` year(4) DEFAULT NULL,
  KEY `expression_index` ((adddate(`a`, interval 10 microsecond)))
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin

insert into t3 values ('2021-01-02', '2021-03-30 08:10:00', '12:01:03', UTC_TIMESTAMP(), 2021);
select * from t3 use index(expression_index) where ADDDATE(a, interval 10 MICROSECOND) = ADDDATE('2021-01-02', interval 10 MICROSECOND);
select * from t3 ignore index(expression_index) where ADDDATE(a, interval 10 MICROSECOND) = ADDDATE('2021-01-02', interval 10 MICROSECOND);
```

### 2. What did you expect to see? (Required)
result should be same when use and ignore index

### 3. What did you see instead (Required)
```sql
MySQL root@127.0.0.1:test> select * from t3 use index(expression_index) where ADDDATE(a, interval 10 MICROSECOND) = ADDDATE('2021-01-02', interval 10 MICROSECOND);
+---+---+---+---+---+
| a | b | c | d | e |
+---+---+---+---+---+

0 rows in set
Time: 0.006s
MySQL root@127.0.0.1:test> select * from t3 ignore index(expression_index) where ADDDATE(a, interval 10 MICROSECOND) = ADDDATE('2021-01-02', interval 10 MICROSECOND);
+------------+---------------------+----------+---------------------+------+
| a          | b                   | c        | d                   | e    |
+------------+---------------------+----------+---------------------+------+
| 2021-01-02 | 2021-03-30 08:10:00 | 12:01:03 | 2021-08-12 02:27:35 | 2021 |
+------------+---------------------+----------+---------------------+------+


```
### 4. What is your TiDB version? (Required)
```sql
MySQL root@127.0.0.1:test> select tidb_version()\G
***************************[ 1. row ]***************************
tidb_version() | Release Version: v5.2.0-alpha-572-g6b8c15301
Edition: Community
Git Commit Hash: 6b8c1530164be17486935b3be250c8aab6072e73
Git Branch: master
UTC Build Time: 2021-08-10 08:52:58
GoVersion: go1.16.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Issue Number: close #27159 

Problem Summary:
If the decimal is unspecified, we should use the maxFsp, which is compatible with what we do for function.

### What is changed and how it works?

If the decimal is unspecified, we should use the maxFsp, which is compatible with what we do for function.

### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
 'None' 
```

### CODE REVIEW

**Comment by morgo** [comment]
It looks like this test is related:
```

[2021-08-12T14:19:49.458Z] FAIL: convert_test.go:590: testTypeConvertSuite.TestConvert

[2021-08-12T14:19:49.458Z] 

[2021-08-12T14:19:49.458Z] convert_test.go:744:

[2021-08-12T14:19:49.458Z]     signedAccept(c, mysql.TypeDatetime, NewDecFromFloatForTest(20010101100000.123456), "2001-01-01 10:00:00")

[2021-08-12T14:19:49.458Z] convert_test.go:548:

[2021-08-12T14:19:49.458Z]     c.Assert(str, Equals, expected)

[2021-08-12T14:19:49.458Z] ... obtained string = "2001-01-01 10:00:00.125000"

[2021-08-12T14:19:49.458Z] ... expected string = "2001-01-01 10:00:00"
```

**Comment by wjhuang2016** [comment]
> It looks like this test is related:
> 
> ```
> 
> [2021-08-12T14:19:49.458Z] FAIL: convert_test.go:590: testTypeConvertSuite.TestConvert
> 
> [2021-08-12T14:19:49.458Z] 
> 
> [2021-08-12T14:19:49.458Z] convert_test.go:744:
> 
> [2021-08-12T14:19:49.458Z]     signedAccept(c, mysql.TypeDatetime, NewDecFromFloatForTest(20010101100000.123456), "2001-01-01 10:00:00")
> 
> [2021-08-12T14:19:49.458Z] convert_test.go:548:
> 
> [2021-08-12T14:19:49.458Z]     c.Assert(str, Equals, expected)
> 
> [2021-08-12T14:19:49.458Z] ... obtained string = "2001-01-01 10:00:00.125000"
> 
> [2021-08-12T14:19:49.458Z] ... expected string = "2001-01-01 10:00:00"
> ```

Yeah, I am thinking about whether set the decimal to 0 or just change the result.

File: types/datum.go [hunk 1]
```diff
@@ -1122,7 +1122,7 @@ func (d *Datum) convertToMysqlTimestamp(sc *stmtctx.StatementContext, target *Fi
 		t   Time
 		err error
 	)
-	fsp := DefaultFsp
+	fsp := MaxFsp
```
Comment (AilinKid) [inline types/datum.go:1125]: this change will cause some automated-tests fail, be cautious with that since no CI test for now

**Review by AilinKid — APPROVED** [review]
Rest LGTM

**Comment by tangenta** [comment]
Why do you write integration tests in the `explain_test`? :confused:

**Comment by wjhuang2016** [comment]
> Why do you write integration tests in the `explain_test`? 😕

Let's forget its name. We should take it as mysql_test

**Comment by tangenta** [comment]
I can't forget its name :)

**Comment by wjhuang2016** [comment]
> I can't forget its name :)

It's suggested by @zz-jason that we should add the tests into `mysql-test` instead of `integration_test`, I think it's a good idea. You can refer to cte.t and select.t in this folder.

**Review by AilinKid — COMMENTED** [review]
/LGTM

**Review by tangenta — APPROVED** [review]

**Pull request merged**